File: linked_list/2_add_two_numbers.cc
// 2024/01/07: 不看答案自己写出来了.
struct ListNode {
  int val;
  ListNode* next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(next) {}
};

ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
  int carry = 0;
  ListNode* dummy = new ListNode();
  ListNode* curr = dummy;
  while (l1 || l2) {
    int sum = (l1 ? l1->val : 0) + (l2 ? l2->val : 0) + carry;
    carry = sum / 10;
    sum = sum % 10;
    curr->next = new ListNode(sum);
    curr = curr->next;
    if (l1) {
      l1 = l1->next;
    }
    if (l2) {
      l2 = l2->next;
    }
  }
  if (carry == 1) {
    curr->next = new ListNode(carry);
  }
  return dummy->next;
}

File: linked_list/2_add_two_numbers.py
from linked_list.list_node import ListNode, create_list_node, print_list_node
from typing import Optional


# 2026/05/14: 不看答案自己写出来了.
def add_two_numbers(
    l1: Optional[ListNode], l2: Optional[ListNode]
) -> Optional[ListNode]:
    dummy = ListNode(0)
    prev = dummy
    carry = 0
    while l1 and l2:
        num = (l1.val + l2.val + carry) % 10
        carry = (l1.val + l2.val + carry) // 10
        prev.next = ListNode(num)
        prev = prev.next
        l1 = l1.next
        l2 = l2.next

    while l1:
        num = (l1.val + carry) % 10
        carry = (l1.val + carry) // 10
        prev.next = ListNode(num)
        prev = prev.next
        l1 = l1.next

    while l2:
        num = (l2.val + carry) % 10
        carry = (l2.val + carry) // 10
        prev.next = ListNode(num)
        prev = prev.next
        l2 = l2.next

    if carry == 1:
        prev.next = ListNode(carry)

    return dummy.next


l1 = [9, 9, 9, 9, 9, 9, 9]
l2 = [9, 9, 9, 9]
l1 = create_list_node(l1)
l2 = create_list_node(l2)
print_list_node(l1)
print_list_node(l2)
print_list_node(add_two_numbers(l1, l2))
